llvm/lib/Target/AIE/AIEBaseSubtarget.cpp

```diff
@@ -446,17 +446,39 @@ class WAWEdges : public ScheduleDAGMutation {
       }
     }
   }
-  void apply(ScheduleDAGInstrs *DAG) override {
+
+public:
+  void apply(ScheduleDAGInstrs *DAG, AIEPostRASchedStrategy *Scheduler) {
     MachineFunction &MF = DAG->MF;
     MachineRegisterInfo &MRI = MF.getRegInfo();
     const TargetRegisterInfo *TRI = MF.getSubtarget().getRegisterInfo();
     auto *RI = static_cast<const AIEBaseRegisterInfo *>(TRI);
     LivePhysRegs LiveRegs;
     LiveRegs.init(*TRI);
-    // Reserved registers are considered always live
-    for (MCPhysReg PhysReg : MRI.getReservedRegs().set_bits()) {
-      if (RI->isSimplifiableReservedReg(PhysReg))
-        LiveRegs.addReg(PhysReg);
+    if (Scheduler) {
+      MachineBasicBlock *MBB = DAG->getBB();
+      const BlockState &BS = Scheduler->getInterBlock().getBlockState(MBB);
+      auto Region = BS.getCurrentRegion();
+      auto BottomRegion = BS.getBottom();
+      if (*Region.begin() == *BottomRegion.begin()) {
+        // If the region is bottom region, liveouts of region are same as
+        // liveouts of the MBB
+        for (MCPhysReg Reg : BS.LiveOuts) {
+          LiveRegs.addReg(Reg);
+        }
+      } else {
+        // Reserved registers are considered always live
+        for (MCPhysReg PhysReg : MRI.getReservedRegs().set_bits()) {
```

Comment:
We can simplify a bit to avoid the same code con two different places, like:

```
    bool AddReservedRegs = true;
    if (Scheduler) {
      MachineBasicBlock *MBB = DAG->getBB();
      const BlockState &BS = Scheduler->getInterBlock().getBlockState(MBB);
      auto Region = BS.getCurrentRegion();
      auto BottomRegion = BS.getBottom();
      if (*Region.begin() == *BottomRegion.begin()) {
        // If the region is bottom region, liveouts of region are same as
        // liveouts of the MBB
        for (const MCPhysReg Reg : BS.LiveOuts)
          LiveRegs.addReg(Reg);

        AddReservedRegs = false;
      }
    } 

    if(AddReservedRegs) {
      // Reserved registers are considered always live
      for (const MCPhysReg PhysReg : MRI.getReservedRegs().set_bits()) {
        if (RI->isSimplifiableReservedReg(PhysReg))
          LiveRegs.addReg(PhysReg);
      }
    }
```